includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -1098,6 +1117,213 @@ private function sanitize_disallowed_attribute_values_in_node( DOMElement $node,
 		return $attributes_pending_removal;
 	}
 
+	/**
+	 * Ensure the width and height for the element are valid.
+	 *
+	 * @param DOMElement $node Node.
+	 * @return bool True if both width and height are valid, false if either is not valid.
+	 */
+	private function has_valid_dimensions( $node ) {
+		$input_width = new AMP_CSS_Length(
+			$node->getAttribute( 'width' ),
+			true,
+			true
+		);
+
+		if ( ! $input_width->is_valid() ) {
+			return false;
+		}
+
+		$input_height = new AMP_CSS_Length(
+			$node->getAttribute( 'height' ),
+			true,
+			true
+		);
+
+		if ( ! $input_height->is_valid() ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Validates the layout for the given tag. This involves checking the
+	 * layout, width, height, sizes attributes with AMP specific logic.
+	 *
+	 * @param array[][]  $tag_spec Tag spec list.
+	 * @param DOMElement $node     Tag to validate.
+	 * @return bool True if layout is valid, false if validation fails.
+	 */
+	private function is_valid_layout( $tag_spec, $node ) {
+		$layout_attr = $node->getAttribute( 'layout' );
+
+		$input_width = new AMP_CSS_Length(
+			$node->getAttribute( 'width' ),
+			true,
+			AMP_Rule_Spec::LAYOUT_FLUID === $layout_attr
+		);
+
+		if ( ! $input_width->is_valid() ) {
+			return false;
+		}
+
+		$input_height = new AMP_CSS_Length(
+			$node->getAttribute( 'height' ),
+			true,
+			AMP_Rule_Spec::LAYOUT_FLUID === $layout_attr
+		);
+
+		if ( ! $input_height->is_valid() ) {
+			return false;
+		}
+
+		if ( isset( $tag_spec['amp_layout'] ) ) {
```

Comment:
The JS validates starts off its `validateLayout()` function with this check as an assert: https://github.com/ampproject/amphtml/blob/b203bbac123337a83752083ac6018fab92393f9f/validator/engine/validator.js#L3940

That means it wouldn't even calculate the `$input_width` and `$input_height` when the `'amp_layout'` spec is not set, so this deviates from the JS validator behavior.

Was there a specific reason why you moved this check to a later point in the code?

Reply:
Ah, I see what you mean now. That assertion is not meant to be placed at a later point in the `is_valid_layout` method, but instead when checking if [the element has the layout attribute set](https://github.com/ampproject/amp-wp/pull/3728/files#diff-ee41b32e64c90445a71328ca93cfa799R1026).